plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStoreBootstrapper.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugin.ingest;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+
+public final class PipelineStoreBootstrapper extends AbstractLifecycleComponent implements ClusterStateListener {
+
+    private final ThreadPool threadPool;
+    private final PipelineStore pipelineStore;
+    private final ClusterService clusterService;
+    private final TimeValue pipelineUpdateInterval;
+
+    @Inject
+    public PipelineStoreBootstrapper(Settings settings, ThreadPool threadPool, PipelineStore pipelineStore, ClusterService clusterService) {
+        super(settings);
+        this.threadPool = threadPool;
+        this.pipelineStore = pipelineStore;
+        this.clusterService = clusterService;
+        this.pipelineUpdateInterval = settings.getAsTime("ingest.pipeline.store.update.interval", TimeValue.timeValueMinutes(30));
+        clusterService.add(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK) == false) {
+            boolean canInitialize;
+            if (event.state().getMetaData().hasConcreteIndex(PipelineStore.INDEX) == false) {
+                // it will be ready when auto create index kicks in before the first pipeline doc gets added
+                canInitialize = true;
+            } else {
+                IndexRoutingTable routingTable = event.state().getRoutingTable().index(PipelineStore.INDEX);
+                canInitialize = routingTable.allPrimaryShardsActive();
+            }
+
+            if (canInitialize) {
+                pipelineStore.initialize();
+                scheduleBackgroundUpdater();
+                clusterService.remove(this);
+            }
+        }
+    }
+
+    @Override
+    protected void doStart() {
+    }
+
+    @Override
+    protected void doStop() {
+    }
+
+    @Override
+    protected void doClose() {
+        try {
+            pipelineStore.close();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    void scheduleBackgroundUpdater() {
```

Comment:
Let me think... when a node loads the pipelines via scroll search it is already part of the cluster. What I'm concerned about is visibility of pipelines changes in the `.ingest` index. So a node is scroll searching for pipelines, an update is made to a pipeline on some other node. On the node that is loading pipelines both of these actions go through the `PipelineStore#updatePipelines()` method. This method is synchronized, so in this case the reload operation caused by a pipeline update will have to wait until the initial pipeline loading has completed. So my conclusion is that this is safe and we can remove the scheduling of background update tasks.

Does my reasoning make sense?